internal/balancer/balancer_test.go

```diff
@@ -0,0 +1,98 @@
+package balancer
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/test/bufconn"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/config"
+)
+
+func TestBalancer_discoveryConn(t *testing.T) {
+	// testTimeout defines the test timeout and is an example of an actual user-defined timeout.
+	//
+	// I couldn't find any events for synchronization, assuming that there
+	// might be retries with different logic and their own timeouts inside `discoveryConn`.
+	// If not now, then in the future. One second excludes false test failures in case
+	// the test is run on very slow workers. Moreover, one second is only lost if the test fails,
+	// and in that case, losing it is not critical. Upon successful completion of the test,
+	// the context is canceled via `cancel()`.
+	const testTimeout = 1 * time.Second
+
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	fakeListener := bufconn.Listen(1024 * 1024)
+	defer fakeListener.Close()
+
+	fakeServer := grpc.NewServer()
+	defer fakeServer.Stop()
+
+	go func() {
+		_ = fakeServer.Serve(fakeListener)
+	}()
+
+	var dialAttempt uint32
+
+	balancer := &Balancer{
+		address: "ydbmock:///mock",
+		driverConfig: config.New(
+			config.WithEndpoint("mock"),
+			config.WithGrpcOptions(
+				grpc.WithResolvers(&mockResolverBuilder{}),
+
+				grpc.WithContextDialer(
+					// The first dialing is never ended, while the subsequent ones work fine.
+					func(ctx context.Context, s string) (net.Conn, error) {
+						atomic.AddUint32(&dialAttempt, 1)
```

Comment:
What about store returned result, for exactly one time make infinite dial?

for current code two parallel goroutines can increase the variable before atomic.Load read result. And no of goroutines will be infinite.